torch/_decomp/decompositions.py

```diff
@@ -1376,3 +1376,88 @@ def upsample_bilinear2d_vec(
     q2 = torch.mul(v3, xscale1) + torch.mul(v4, xscale2)
     result = torch.mul(q1, yscale1) + torch.mul(q2, yscale2)
     return result
+
+
+# We should be applying decompositions after all transformations
+@register_decomposition(aten.is_same_size.default)
+def is_same_size(a: Tensor, b: Tensor) -> bool:
+    return a.shape == b.shape
+
+
+@register_decomposition(aten.nll_loss_forward)
+def nll_loss_forward(
+    self: Tensor,
+    target: Tensor,
+    weight: Optional[Tensor],
+    reduction: int,
+    ignore_index: int,
+) -> Tuple[Tensor, Tensor]:
+    assert self.dim() > 0 and self.dim() <= 2, "input tensor should be 1D or 2D"
+    assert (
+        target.dim() <= 1
+    ), "0D or 1D target tensor expected, multi-target not supported"
+
+    no_batch_dim = self.dim() == 1 and target.dim() == 0
+    assert no_batch_dim or (
+        self.shape[0] == target.shape[0]
+    ), f"size mismatch (got input: {self.shape}, target: {target.shape})"
+
+    n_classes = self.shape[-1]
+
+    assert weight is None or (
+        weight.dim() == 1 and weight.numel() == n_classes
+    ), f"weight tensor should be defined either for all {n_classes} classes or no classes but got weight tensor of shape: {weight.shape}"  # noqa: B950
+
+    # self can be [N, C] or [C]
+    # target can be [N] or []
+
+    n_dims = self.dim()
+    channel_dim = 1
+    if n_dims < 2:
+        channel_dim = 0
+
+    if weight is not None:
+        # Here is a specific case with reduction mean and non-batched tensors
+        # https://github.com/pytorch/pytorch/issues/61309
+        # In this case weight is cancelled: w * x[t] / w -> x[t]
+        if not (reduction == Reduction.MEAN.value and n_dims < 2):
+            w = weight.unsqueeze(0) if n_dims > 1 else weight
+            self = self * w
+
+    target_ = target.unsqueeze(channel_dim)
+    # target can be [N, 1] or [1]
+
+    result = -torch.gather(self, channel_dim, target_).squeeze(channel_dim)
+
+    ignore_index_mask = None
+    if ignore_index >= 0:
+        ignore_index_mask = target != ignore_index
+        result = result * ignore_index_mask
```

Comment:
what about spurious nan propagation here, or does it happen the same way in eager?

Reply:
iirc, `target` has to be an integer tensor, so it's not possible to have nans in `ignore_index_mask`?

Reply:
But it's possible to have nans in result that won't be masked by multiplying by zero?

Reply:
Yeah, you're right, I think eager does the "right" thing.